### PR TITLE
fix(site): Close WS and SE connections when they are not used

### LINKED ISSUE #6782: Websocket connections are not closing when users navigating away from the page
@Kira-Pilot found this for the watchBuildLogs state in the template editor page but this is probably happening for all the other websocket connections that we have in the site.

### PR DESCRIPTION
**Pull request by BrunoQuaresma**
About the fix:
- Change the watch services to return a function closing the connection
- Refactor the promises to regular functions where they can do actions when receiving a message, or error or when it is done
- Transition the watching state to something else using a callback event

⚠️ This touches some important parts of the UI so I would love some extra QA:
- Workspace page - Check if the statuses are being updated automatically and the startup logs as well
- Template version logs  - Check if the logs are being displayed correctly
- Template version editor - Check if the logs are showing correctly

Fix #6782

### CODE REVIEW

File: site/src/api/api.ts [hunk 1]
```diff
@@ -1106,25 +1106,107 @@ export const watchAgentMetadata = (agentId: string): EventSource => {
...
+  })
+  return socket
+}
+
+type WatchStartupLogsOptions = {
+  after: number
+  onMessage: (logs: TypesGen.WorkspaceAgentStartupLog[]) => void
+  onDone: () => void
+  onError: (error: Error) => void
+}
```
Comment (Kira-Pilot) [inline site/src/api/api.ts:1154]: Can we just use one type that takes a generic parameter for logs type?

**Review by Kira-Pilot — APPROVED** [review]
This looks like its working to me!

File: site/src/xServices/templateVersionEditor/templateVersionEditorXService.ts [hunk 2]
```diff
@@ -360,9 +359,21 @@ export const templateVersionEditorMachine = createMachine(
...
+              callback({ type: "BUILD_DONE" })
+            },
+            onError: (error) => {
+              console.error(error)
+            },
           })
+
+          return () => {
+            socket.close()
+          }
```
[the remark this answers is not in the document]
Comment (BrunoQuaresma) [reply site/src/xServices/templateVersionEditor/templateVersionEditorXService.ts:376]: exactly!

File: site/src/api/api.ts (hunk 1 above)
Comment (BrunoQuaresma) [reply site/src/api/api.ts:1154]: Yes, we can!

**Pull request merged**